examples/bigquery-row-access-groups/group_sync.py

```diff
@@ -0,0 +1,233 @@
+# the License.
+#
+# Any software provided by Google hereunder is distributed "AS IS", WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, and is not intended for production use.
+"""Creates a BigQuery table listing members of all groups in a GSuite domain."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import argparse
+import csv
+from io import StringIO
+import sys
+
+import auth_util
+from apiclient.discovery import build
+import google.api_core.exceptions
+from google.cloud import bigquery
+
+
+class GroupSync(object):
+  """Manages synchronization of group member data from GSuite API to BigQuery."""
+
+  def __init__(self, admin_email, domain):
+    self.bq_client = bigquery.Client()
+    self.service = self._get_service_reference(admin_email)
+    self.domain = domain
+
+  def _get_service_reference(self, admin_email):
+    """Instantiates a connection to the GSuite API.
+
+    Args:
+      admin_email: a GSuite administrator account to impersonate.
+
+    Returns:
+      A reference to an API Service object, ready to call GSuite APIs.
+    """
+    scopes = [
+        'https://www.googleapis.com/auth/admin.directory.group',
+        # For Cloud Identity, replace with the scope below
+        # 'https://www.googleapis.com/auth/cloud-identity.groups.readonly'
+    ]
+    credentials, _ = auth_util.get_credentials(admin_email, scopes)
+    service = build('admin', 'directory_v1', credentials=credentials)
+    # For Cloud Identity, replace with the statement below
+    # service = build('cloudidentity', 'v1', credentials=credentials)
+    return service
+
+  def get_group_members_as_list(self):
+    """Retrieves the members of all groups in the domain as an array.
+
+    Returns:
+      A bi-dimensional array with each row representing a pair of group id and
+      user id (both strings).
+    """
+
+    all_group_members = []
+    results = self.service.groups().list(domain=self.domain).execute()
+    # For Cloud Identity, replace with the statement below
+    # results = self.service.groups().list(parent=self.domain).execute()
+
+    for g in results['groups']:
+      group_id = g['email']
+      members = self._list_group_members(group_id)
+      group_members = zip([group_id] * len(members), members)
+      all_group_members.extend(group_members)
+
+    return all_group_members
+
+  def get_group_members_as_map(self):
+    """Retrieves the members of all groups in the domain as an array.
+
+    Returns:
+      A map in which the keys are group ids and the values are arrays of user
+      ids belonging to that group.
+    """
+
+    group_map = {}
+    results = self.service.groups().list(domain=self.domain).execute()
+
+    for g in results['groups']:
+      group_id = g['email']
+      members = self._list_group_members(group_id)
+      group_map[group_id] = members
+
+    return group_map
+
+  def _list_group_members(self, group_id):
```

Comment:
For the sake of having unit tests, please add a unit test for this function. Might need to accept the output of `service.members().list(groupKey=groupId).execute` to ease mocking for unit testing.

Reply:
@cbreuel please add a unit test for this function for constructing this list of group members structure. Consider reparameterizing so this function accepts the output of the list member api call for ease of testing.

Reply:
Done.